File: fhir-bridge/dotnet/src/FhirBridge/PedigreeDetail.cs
using System.Text.Json;

namespace FhirBridge;

public sealed class PedigreeParseException : Exception
{
    public PedigreeParseException(string message) : base(message) { }
}

public sealed record PedigreeIndividual(
    string Id,
    string DisplayName,
    BiologicalSex BiologicalSex,
    bool Proband,
    int? YearOfBirth,
    string? BornOn);

public sealed record PedigreeRelationship(string Id, IReadOnlyList<string> Members);

public sealed record PedigreeEgg(string IndividualId, string RelationshipId);

public sealed record PedigreeDetail(
    string Id,
    string DisplayName,
    IReadOnlyList<PedigreeIndividual> Individuals,
    IReadOnlyList<PedigreeRelationship> Relationships,
    IReadOnlyList<PedigreeEgg> Eggs);

public static class PedigreeDetailParser
{
    public static PedigreeDetail Parse(JsonElement root)
    {
        if (root.ValueKind != JsonValueKind.Object)
        {
            throw new PedigreeParseException("pedigree response is not a JSON object");
        }
        return new PedigreeDetail(
            Id: RequireString(root, "id"),
            DisplayName: RequireString(root, "display_name"),
            Individuals: RequireArray(root, "individuals").Select(ParseIndividual).ToList(),
            Relationships: RequireArray(root, "relationships").Select(ParseRelationship).ToList(),
            Eggs: RequireArray(root, "eggs").Select(ParseEgg).ToList());
    }

    private static PedigreeIndividual ParseIndividual(JsonElement element)
    {
        RequireObject(element, "individual");
        var properties = OptionalObject(element, "properties");
        int? year = null;
        if (properties is { } props && props.TryGetProperty("year_of_birth", out var yearElement) &&
            yearElement.ValueKind == JsonValueKind.Number && yearElement.TryGetInt32(out var y))
        {
            year = y;
        }
        return new PedigreeIndividual(
            Id: RequireString(element, "id"),
            DisplayName: RequireString(element, "display_name"),
            BiologicalSex: BiologicalSexWire.Parse(RequireString(element, "biological_sex")),
            Proband: ReadProbandFlag(element),
            YearOfBirth: year,
            BornOn: EarliestBirthDate(element));
    }

    private static PedigreeRelationship ParseRelationship(JsonElement element)
    {
        RequireObject(element, "relationship");
        var members = new List<string>();
        foreach (var member in RequireArray(element, "members"))
        {
            if (member.ValueKind != JsonValueKind.String)
            {
                throw new PedigreeParseException("relationship.members entries must be strings");
            }
            members.Add(member.GetString() ?? string.Empty);
        }
        return new PedigreeRelationship(RequireString(element, "id"), members);
    }

    private static PedigreeEgg ParseEgg(JsonElement element)
    {
        RequireObject(element, "egg");
        return new PedigreeEgg(
            IndividualId: RequireString(element, "individual_id"),
            RelationshipId: RequireString(element, "relationship_id"));
    }

    private static bool ReadProbandFlag(JsonElement element)
    {
        if (!element.TryGetProperty("proband", out var flag))
        {
            return false;
        }
        return flag.ValueKind switch
        {
            JsonValueKind.Number when flag.TryGetInt32(out var value) => value == 1,
            JsonValueKind.True => true,
            _ => false,
        };
    }

    private static string? EarliestBirthDate(JsonElement element)
    {
        if (!element.TryGetProperty("events", out var events) || events.ValueKind != JsonValueKind.Array)
        {
            return null;
        }
        foreach (var ev in events.EnumerateArray())
        {
            if (ev.ValueKind != JsonValueKind.Object)
            {
                continue;
            }
            if (ev.TryGetProperty("type", out var type) && type.GetString() == "birth" &&
                ev.TryGetProperty("date_start", out var date) && date.ValueKind == JsonValueKind.String)
            {
                return date.GetString();
            }
        }
        return null;
    }

    private static void RequireObject(JsonElement element, string label)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            throw new PedigreeParseException($"Expected object for {label}");
        }
    }

    private static string RequireString(JsonElement element, string key)
    {
        if (!element.TryGetProperty(key, out var value) || value.ValueKind != JsonValueKind.String)
        {
            throw new PedigreeParseException($"Missing string field '{key}'");
        }
        return value.GetString() ?? throw new PedigreeParseException($"Field '{key}' is null");
    }

    private static IEnumerable<JsonElement> RequireArray(JsonElement element, string key)
    {
        if (!element.TryGetProperty(key, out var value) || value.ValueKind != JsonValueKind.Array)
        {
            throw new PedigreeParseException($"Missing array field '{key}'");
        }
        foreach (var item in value.EnumerateArray())
        {
            yield return item;
        }
    }

    private static JsonElement? OptionalObject(JsonElement element, string key)
    {
        if (element.TryGetProperty(key, out var value) && value.ValueKind == JsonValueKind.Object)
        {
            return value;
        }
        return null;
    }
}

File: fhir-bridge/node/src/pedigreeDetail.ts
/**
 * Narrow, read-only projection of the Evagene /api/pedigrees/{id}
 * response. Only the fields the bridge actually uses.
 */

export type PedigreeBiologicalSex = 'female' | 'male' | 'unknown';

export interface PedigreeIndividual {
  readonly id: string;
  readonly displayName: string;
  readonly biologicalSex: PedigreeBiologicalSex;
  readonly proband: boolean;
  readonly yearOfBirth?: number;
  readonly bornOn?: string;
}

export interface PedigreeRelationship {
  readonly id: string;
  readonly members: readonly string[];
}

export interface PedigreeEgg {
  readonly individualId: string;
  readonly relationshipId: string;
}

export interface PedigreeDetail {
  readonly id: string;
  readonly displayName: string;
  readonly individuals: readonly PedigreeIndividual[];
  readonly relationships: readonly PedigreeRelationship[];
  readonly eggs: readonly PedigreeEgg[];
}

export class PedigreeParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'PedigreeParseError';
  }
}

export function parsePedigreeDetail(raw: unknown): PedigreeDetail {
  const root = requireObject(raw, 'pedigree');
  return {
    id: requireString(root, 'id'),
    displayName: requireString(root, 'display_name'),
    individuals: requireArray(root, 'individuals').map(parseIndividual),
    relationships: requireArray(root, 'relationships').map(parseRelationship),
    eggs: requireArray(root, 'eggs').map(parseEgg),
  };
}

function parseIndividual(raw: unknown): PedigreeIndividual {
  const obj = requireObject(raw, 'individual');
  const props = optionalObject(obj, 'properties') ?? {};
  const year = optionalNumber(props, 'year_of_birth');
  const bornOn = earliestBirthDate(optionalArray(obj, 'events') ?? []);
  return {
    id: requireString(obj, 'id'),
    displayName: requireString(obj, 'display_name'),
    biologicalSex: parseBiologicalSex(requireString(obj, 'biological_sex')),
    proband: parseProbandFlag(obj.proband),
    ...(year === undefined ? {} : { yearOfBirth: year }),
    ...(bornOn === undefined ? {} : { bornOn }),
  };
}

function parseRelationship(raw: unknown): PedigreeRelationship {
  const obj = requireObject(raw, 'relationship');
  return {
    id: requireString(obj, 'id'),
    members: requireArray(obj, 'members').map(m => {
      if (typeof m !== 'string') {
        throw new PedigreeParseError('relationship.members entries must be strings');
      }
      return m;
    }),
  };
}

function parseEgg(raw: unknown): PedigreeEgg {
  const obj = requireObject(raw, 'egg');
  return {
    individualId: requireString(obj, 'individual_id'),
    relationshipId: requireString(obj, 'relationship_id'),
  };
}

function parseBiologicalSex(raw: string): PedigreeBiologicalSex {
  if (raw === 'female' || raw === 'male' || raw === 'unknown') {
    return raw;
  }
  return 'unknown';
}

function parseProbandFlag(raw: unknown): boolean {
  if (raw === 1 || raw === true) {
    return true;
  }
  return false;
}

function earliestBirthDate(events: readonly unknown[]): string | undefined {
  for (const event of events) {
    if (typeof event !== 'object' || event === null) {
      continue;
    }
    const record = event as Record<string, unknown>;
    if (record.type === 'birth' && typeof record.date_start === 'string') {
      return record.date_start;
    }
  }
  return undefined;
}

function requireObject(raw: unknown, label: string): Record<string, unknown> {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    throw new PedigreeParseError(`Expected object for ${label}`);
  }
  return raw as Record<string, unknown>;
}

function requireString(obj: Record<string, unknown>, key: string): string {
  const value = obj[key];
  if (typeof value !== 'string') {
    throw new PedigreeParseError(`Missing string field '${key}'`);
  }
  return value;
}

function requireArray(obj: Record<string, unknown>, key: string): readonly unknown[] {
  const value = obj[key];
  if (!Array.isArray(value)) {
    throw new PedigreeParseError(`Missing array field '${key}'`);
  }
  return value;
}

function optionalArray(obj: Record<string, unknown>, key: string): readonly unknown[] | undefined {
  const value = obj[key];
  return Array.isArray(value) ? value : undefined;
}

function optionalObject(
  obj: Record<string, unknown>,
  key: string,
): Record<string, unknown> | undefined {
  const value = obj[key];
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return undefined;
  }
  return value as Record<string, unknown>;
}

function optionalNumber(obj: Record<string, unknown>, key: string): number | undefined {
  const value = obj[key];
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined;
}
